In [2]:
import pandas as pd

In [3]:
pd.set_option("display.max_columns", None)

# Data CAPEC mutilable dataset


In [4]:
df_capec = pd.read_csv('Dataset/data_capec_multilabel.csv')

/tmp/ipykernel_51021/1247042814.py:1: DtypeWarning: Columns (9,11,12,13,14,15,16,18,19) have mixed types. Specify dtype option on import or set low_memory=False.
  df_capec = pd.read_csv('Dataset/data_capec_multilabel.csv')


In [5]:
df_capec.head()

timestamp      src_ip  src_port      dst_ip  dst_port  \
0  17/Jul/2020:12:23:34 +0100  172.26.0.1   55894.0  172.26.0.4      80.0   
1  17/Jul/2020:12:23:41 +0100  172.26.0.1   55897.0  172.26.0.4      80.0   
2  17/Jul/2020:12:23:51 +0100  172.26.0.1   55901.0  172.26.0.4      80.0   
3  17/Jul/2020:12:23:51 +0100  172.26.0.1   55902.0  172.26.0.4      80.0   
4  17/Jul/2020:12:23:58 +0100  172.26.0.1   55903.0  172.26.0.4      80.0   

  request_http_method                               request_http_request  \
0                 GET                                                  /   
1                 GET  /blog/index.php/2020/04/04/voluptatum-reprehen...   
2                 GET                               /blog/xmlrpc.php?rsd   
3                 GET                                                  /   
4                 GET  /blog/index.php/2020/04/04/nihil-tenetur-et-ad...   

  request_http_protocol                                 request_user_agent  \
0              HTTP/1.1  Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_3...   
1              HTTP/1.1  Mozilla/5.0 (Macintosh; Intel Mac OS X 10.6; r...   
2              HTTP/1.1  Mozilla/4.0 (compatible; MSIE 6.0; Windows NT ...   
3              HTTP/1.1  Mozilla/5.0 (Linux; Android 4.2.2; SM-T110 Bui...   
4              HTTP/1.1  Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_2...   

  request_referer   request_host request_origin request_cookie  \
0             NaN  test-site.com            NaN            NaN   
1             NaN  test-site.com            NaN            NaN   
2             NaN  test-site.com            NaN            NaN   
3             NaN  test-site.com            NaN            NaN   
4             NaN  test-site.com            NaN            NaN   

  request_content_type request_accept request_accept_language  \
0                  NaN            */*                     NaN   
1                  NaN            */*                     NaN   
2                  NaN            */*                     NaN   
3                  NaN            */*                     NaN   
4                  NaN            */*                     NaN   

  request_accept_encoding  request_do_not_track request_connection  \
0           gzip, deflate                   NaN         keep-alive   
1           gzip, deflate                   NaN         keep-alive   
2           gzip, deflate                   NaN         keep-alive   
3           gzip, deflate                   NaN         keep-alive   
4           gzip, deflate                   NaN         keep-alive   

  request_body response_http_protocol  response_http_status_code  \
0          NaN               HTTP/1.1                      200.0   
1          NaN               HTTP/1.1                      200.0   
2          NaN               HTTP/1.1                      200.0   
3          NaN               HTTP/1.1                      200.0   
4          NaN               HTTP/1.1                      200.0   

  response_http_status_message  response_content_length  000 - Normal  \
0                           OK                  25174.0             1   
1                           OK                  20862.0             1   
2                           OK                    317.0             1   
3                           OK                  25174.0             1   
4                           OK                  20863.0             1   

   272 - Protocol Manipulation  242 - Code Injection  \
0                            0                     0   
1                            0                     0   
2                            0                     0   
3                            0                     0   
4                            0                     0   

   88 - OS Command Injection  126 - Path Traversal  66 - SQL Injection  \
0                          0                     0                   0   
1                          0                     0                   0   
2              

In [6]:
## Delete blank text in http request
df_capec = df_capec[df_capec['request_http_request'] != '/']

## Convert from multlabel to multiclass

In [7]:
df_capec.shape

(608806, 38)

In [8]:
def count_duplicates(a0, a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13):
    return a0+a1+a2+a3+a4+a5+a6+a7+a8+a9+a10+a11+a12+a13
df_capec['count'] = df_capec.apply(lambda x: count_duplicates(x['000 - Normal'],x['272 - Protocol Manipulation'], x['242 - Code Injection'],x['88 - OS Command Injection'],x['126 - Path Traversal'],x['66 - SQL Injection'],x['16 - Dictionary-based Password Attack'],x['310 - Scanning for Vulnerable Software'],x['153 - Input Data Manipulation'],x['248 - Command Injection'],x['274 - HTTP Verb Tampering'],x['194 - Fake the Source of Data'],x['34 - HTTP Response Splitting'],x['33 - HTTP Request Smuggling']), axis = 1)

### Data with only multiclass 

In [9]:
df_multiclass = df_capec[df_capec['count'] == 1]

df_multiclass.shape

(599755, 39)

In [10]:
df_multiclass.head()

timestamp      src_ip  src_port      dst_ip  dst_port  \
1  17/Jul/2020:12:23:41 +0100  172.26.0.1   55897.0  172.26.0.4      80.0   
2  17/Jul/2020:12:23:51 +0100  172.26.0.1   55901.0  172.26.0.4      80.0   
4  17/Jul/2020:12:23:58 +0100  172.26.0.1   55903.0  172.26.0.4      80.0   
5  17/Jul/2020:12:24:07 +0100  172.26.0.1   55910.0  172.26.0.4      80.0   
6  17/Jul/2020:12:24:15 +0100  172.26.0.1   55911.0  172.26.0.4      80.0   

  request_http_method                               request_http_request  \
1                 GET  /blog/index.php/2020/04/04/voluptatum-reprehen...   
2                 GET                               /blog/xmlrpc.php?rsd   
4                 GET  /blog/index.php/2020/04/04/nihil-tenetur-et-ad...   
5                 GET  /blog/index.php/2020/04/04/explicabo-qui-fuga-...   
6                 GET  /blog/index.php/2020/04/04/explicabo-qui-fuga-...   

  request_http_protocol                                 request_user_agent  \
1              HTTP/1.1  Mozilla/5.0 (Macintosh; Intel Mac OS X 10.6; r...   
2              HTTP/1.1  Mozilla/4.0 (compatible; MSIE 6.0; Windows NT ...   
4              HTTP/1.1  Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_2...   
5              HTTP/1.1  Mozilla/5.0 (Windows NT 10.0; Win64; x64) Appl...   
6              HTTP/1.1  Mozilla/5.0 (Windows NT 5.1) AppleWebKit/537.3...   

  request_referer   request_host request_origin request_cookie  \
1             NaN  test-site.com            NaN            NaN   
2             NaN  test-site.com            NaN            NaN   
4             NaN  test-site.com            NaN            NaN   
5             NaN  test-site.com            NaN            NaN   
6             NaN  test-site.com            NaN            NaN   

  request_content_type request_accept request_accept_language  \
1                  NaN            */*                     NaN   
2                  NaN            */*                     NaN   
4                  NaN            */*                     NaN   
5                  NaN            */*                     NaN   
6                  NaN            */*                     NaN   

  request_accept_encoding  request_do_not_track request_connection  \
1           gzip, deflate                   NaN         keep-alive   
2           gzip, deflate                   NaN         keep-alive   
4           gzip, deflate                   NaN         keep-alive   
5           gzip, deflate                   NaN         keep-alive   
6           gzip, deflate                   NaN         keep-alive   

  request_body response_http_protocol  response_http_status_code  \
1          NaN               HTTP/1.1                      200.0   
2          NaN               HTTP/1.1                      200.0   
4          NaN               HTTP/1.1                      200.0   
5          NaN               HTTP/1.1                      200.0   
6          NaN               HTTP/1.1                      200.0   

  response_http_status_message  response_content_length  000 - Normal  \
1                           OK                  20862.0             1   
2                           OK                    317.0             1   
4                           OK                  20863.0             1   
5                           OK                  21062.0             1   
6                           OK                  21061.0             1   

   272 - Protocol Manipulation  242 - Code Injection  \
1                            0                     0   
2                            0                     0   
4                            0                     0   
5                            0                     0   
6                            0                     0   

   88 - OS Command Injection  126 - Path Traversal  66 - SQL Injection  \
1                          0                     0                   0   
2                          0                     0                   0   
4              

### Data with multilabel

In [11]:
df_multilabel = df_capec[df_capec['count'] > 1]
df_multilabel.shape

(9051, 39)

In [12]:
df_multilabel.head()

timestamp      src_ip  src_port      dst_ip  dst_port  \
2940  17/Jul/2020:19:01:39 +0200  172.26.0.1   45500.0  172.26.0.4      80.0   
2961  17/Jul/2020:19:11:23 +0200  172.26.0.1   47482.0  172.26.0.4      80.0   
2969  17/Jul/2020:19:25:15 +0200  172.26.0.1   59862.0  172.26.0.4      80.0   
2974  17/Jul/2020:19:25:16 +0200  172.26.0.1   59862.0  172.26.0.4      80.0   
2977  17/Jul/2020:19:25:16 +0200  172.26.0.1   59862.0  172.26.0.4      80.0   

     request_http_method                               request_http_request  \
2940                POST  /vendor/phpunit/phpunit/src/Util/PHP/eval-stdi...   
2961                POST                               /cgi-bin/ViewLog.asp   
2969                 GET                                        /.svn/wc.db   
2974                 GET                                   /blog/.svn/wc.db   
2977                 GET              /blog/index.php/my-account/.svn/wc.db   

     request_http_protocol                                 request_user_agent  \
2940              HTTP/1.1  Mozilla/5.0 (Windows NT 10.0; Win64; x64) Appl...   
2961              HTTP/1.1                                 B4ckdoor-owned-you   
2969              HTTP/1.1  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
2974              HTTP/1.1  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
2977              HTTP/1.1  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   

     request_referer   request_host request_origin  \
2940             NaN  172.26.0.1:80            NaN   
2961             NaN      127.0.0.1            NaN   
2969             NaN  test-site.com            NaN   
2974             NaN  test-site.com            NaN   
2977             NaN  test-site.com            NaN   

                                         request_cookie  \
2940                                                NaN   
2961                                                NaN   
2969                                                NaN   
2974  wordpress_logged_in_1aefbe2f76edd740f8e362f39d...   
2977  wordpress_logged_in_1aefbe2f76edd740f8e362f39d...   

                   request_content_type request_accept  \
2940  application/x-www-form-urlencoded            NaN   
2961  application/x-www-form-urlencoded            */*   
2969                                NaN            NaN   
2974                                NaN            NaN   
2977                                NaN            NaN   

     request_accept_language request_accept_encoding  request_do_not_track  \
2940                     NaN                    gzip                   NaN   
2961                     NaN           gzip, deflate                   NaN   
2969                     NaN                     NaN                   NaN   
2974                     NaN                     NaN                   NaN   
2977                     NaN                     NaN                   NaN   

     request_connection                                       request_body  \
2940              close                                <?=md5("phpunit")?>   
2961         keep-alive   remote_submit_Flag=1&remote_syslog_Flag=1&Rem...   
2969                NaN                                                NaN   
2974                NaN                                                NaN   
2977                NaN                                                NaN   

     response_http_protocol  response_http_status_code  \
2940               HTTP/1.1                      404.0   
2961               HTTP/1.1                      404.0   
2969               HTTP/1.1                      404.0   
2974               HTTP/1.1                      404.0   
2977               HTTP/1.1                      404.0   

     response_http_status_message  response_content_length  000 - Normal  \
2940                    Not Found                    275.0             0   
2961                    Not Found                    271.0             0   
2969                    Not Found   

In [13]:
#Assign for test only
df_capec = df_multiclass

## 000 - Normal

In [14]:
df_000 = df_capec[df_capec['000 - Normal'] == 1].loc[:, ['request_http_method', 
                                                         'request_http_request',
                                                         'request_body',
                                                         '000 - Normal']].reset_index(drop=True)
df_000.shape

(226509, 4)

In [15]:
df_000 = df_capec[df_capec['000 - Normal'] == 1]
df_000.shape

(226509, 39)

In [16]:
df_000.head()

timestamp      src_ip  src_port      dst_ip  dst_port  \
1  17/Jul/2020:12:23:41 +0100  172.26.0.1   55897.0  172.26.0.4      80.0   
2  17/Jul/2020:12:23:51 +0100  172.26.0.1   55901.0  172.26.0.4      80.0   
4  17/Jul/2020:12:23:58 +0100  172.26.0.1   55903.0  172.26.0.4      80.0   
5  17/Jul/2020:12:24:07 +0100  172.26.0.1   55910.0  172.26.0.4      80.0   
6  17/Jul/2020:12:24:15 +0100  172.26.0.1   55911.0  172.26.0.4      80.0   

  request_http_method                               request_http_request  \
1                 GET  /blog/index.php/2020/04/04/voluptatum-reprehen...   
2                 GET                               /blog/xmlrpc.php?rsd   
4                 GET  /blog/index.php/2020/04/04/nihil-tenetur-et-ad...   
5                 GET  /blog/index.php/2020/04/04/explicabo-qui-fuga-...   
6                 GET  /blog/index.php/2020/04/04/explicabo-qui-fuga-...   

  request_http_protocol                                 request_user_agent  \
1              HTTP/1.1  Mozilla/5.0 (Macintosh; Intel Mac OS X 10.6; r...   
2              HTTP/1.1  Mozilla/4.0 (compatible; MSIE 6.0; Windows NT ...   
4              HTTP/1.1  Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_2...   
5              HTTP/1.1  Mozilla/5.0 (Windows NT 10.0; Win64; x64) Appl...   
6              HTTP/1.1  Mozilla/5.0 (Windows NT 5.1) AppleWebKit/537.3...   

  request_referer   request_host request_origin request_cookie  \
1             NaN  test-site.com            NaN            NaN   
2             NaN  test-site.com            NaN            NaN   
4             NaN  test-site.com            NaN            NaN   
5             NaN  test-site.com            NaN            NaN   
6             NaN  test-site.com            NaN            NaN   

  request_content_type request_accept request_accept_language  \
1                  NaN            */*                     NaN   
2                  NaN            */*                     NaN   
4                  NaN            */*                     NaN   
5                  NaN            */*                     NaN   
6                  NaN            */*                     NaN   

  request_accept_encoding  request_do_not_track request_connection  \
1           gzip, deflate                   NaN         keep-alive   
2           gzip, deflate                   NaN         keep-alive   
4           gzip, deflate                   NaN         keep-alive   
5           gzip, deflate                   NaN         keep-alive   
6           gzip, deflate                   NaN         keep-alive   

  request_body response_http_protocol  response_http_status_code  \
1          NaN               HTTP/1.1                      200.0   
2          NaN               HTTP/1.1                      200.0   
4          NaN               HTTP/1.1                      200.0   
5          NaN               HTTP/1.1                      200.0   
6          NaN               HTTP/1.1                      200.0   

  response_http_status_message  response_content_length  000 - Normal  \
1                           OK                  20862.0             1   
2                           OK                    317.0             1   
4                           OK                  20863.0             1   
5                           OK                  21062.0             1   
6                           OK                  21061.0             1   

   272 - Protocol Manipulation  242 - Code Injection  \
1                            0                     0   
2                            0                     0   
4                            0                     0   
5                            0                     0   
6                            0                     0   

   88 - OS Command Injection  126 - Path Traversal  66 - SQL Injection  \
1                          0                     0                   0   
2                          0                     0                   0   
4              

## 272 - Protocol Manipulation

In [17]:
df_272 = df_capec[df_capec['272 - Protocol Manipulation'] == 1].loc[:, ['request_http_method',
                                                                        'request_http_request', 
                                                                        'request_body', 
                                                                        '272 - Protocol Manipulation']].reset_index(drop=True)
df_272.shape

(6924, 4)

In [18]:
df_272 = df_capec[df_capec['272 - Protocol Manipulation']==1]
df_272.shape

(6924, 39)

In [19]:
df_272.head()

timestamp      src_ip  src_port      dst_ip  dst_port  \
3461  17/Jul/2020:19:30:37 +0200  172.26.0.1   59927.0  172.26.0.4      80.0   
7138  17/Jul/2020:19:58:39 +0200  172.26.0.1   63015.0  172.26.0.4      80.0   
7139  17/Jul/2020:19:58:39 +0200  172.26.0.1   63016.0  172.26.0.4      80.0   
7508  17/Jul/2020:20:09:38 +0200  172.26.0.1   64214.0  172.26.0.4      80.0   
7509  17/Jul/2020:20:09:38 +0200  172.26.0.1   64213.0  172.26.0.4      80.0   

     request_http_method                               request_http_request  \
3461                 GET                        /blog/index.php/my-account/   
7138                 GET  /blog/wp-content/uploads/2020/04/c874f043-4188...   
7139                 GET  /blog/wp-content/uploads/2020/04/5d0beb80-cdb9...   
7508                 GET  /blog/wp-content/uploads/2020/04/8427d570-74b5...   
7509                 GET  /blog/wp-content/uploads/2020/04/cda5c6a3-e3d9...   

     request_http_protocol                                 request_user_agent  \
3461              HTTP/1.1  Mozilla/5.0 (X11; Linux x86_64; rv:73.0) Gecko...   
7138              HTTP/1.1  Mozilla/5.0 (X11; Linux x86_64; rv:73.0) Gecko...   
7139              HTTP/1.1  Mozilla/5.0 (X11; Linux x86_64; rv:73.0) Gecko...   
7508              HTTP/1.1  Mozilla/5.0 (X11; Linux x86_64; rv:73.0) Gecko...   
7509              HTTP/1.1  Mozilla/5.0 (X11; Linux x86_64; rv:73.0) Gecko...   

            request_referer request_host request_origin request_cookie  \
3461                    NaN   172.26.0.1            NaN            NaN   
7138                    NaN   172.26.0.1            NaN            NaN   
7139                    NaN   172.26.0.1            NaN            NaN   
7508  http://test-site.com/   172.26.0.1            NaN            NaN   
7509  http://test-site.com/   172.26.0.1            NaN            NaN   

     request_content_type                                     request_accept  \
3461                  NaN  text/html,application/xhtml+xml,application/xm...   
7138                  NaN                                     image/webp,*/*   
7139                  NaN                                     image/webp,*/*   
7508                  NaN                                     image/webp,*/*   
7509                  NaN                                     image/webp,*/*   

     request_accept_language request_accept_encoding  request_do_not_track  \
3461          en-US,en;q=0.5                     NaN                   1.0   
7138          en-US,en;q=0.5                     NaN                   1.0   
7139          en-US,en;q=0.5                     NaN                   1.0   
7508          en-US,en;q=0.5           gzip, deflate                   1.0   
7509          en-US,en;q=0.5           gzip, deflate                   1.0   

     request_connection request_body response_http_protocol  \
3461         keep-alive          NaN               HTTP/1.1   
7138         keep-alive          NaN               HTTP/1.1   
7139         keep-alive          NaN               HTTP/1.1   
7508         keep-alive          NaN               HTTP/1.1   
7509         keep-alive          NaN               HTTP/1.1   

      response_http_status_code response_http_status_message  \
3461                      200.0                           OK   
7138                      200.0                           OK   
7139                      200.0                           OK   
7508                      200.0                           OK   
7509                      200.0                           OK   

      response_content_length  000 - Normal  272 - Protocol Manipulation  \
3461                  57822.0             0                            1   
7138                  78796.0             0                            1   
7139                 236440.0             0                            1   
7508                   2255.0             0                            1   
7509                   1461.0            

## 242 - Code Injection

In [20]:
df_242 = df_capec[df_capec['242 - Code Injection'] == 1].loc[:, ['request_http_method',
                                                                        'request_http_request', 
                                                                        'request_body', 
                                                                        '242 - Code Injection']].reset_index(drop=True)
df_242.shape

(13792, 4)

In [21]:
df_242 = df_capec[df_capec['242 - Code Injection'] == 1]
df_242.shape

(13792, 39)

In [22]:
df_242.head()

timestamp      src_ip  src_port      dst_ip  \
7491    17/Jul/2020:20:08:51 +0200  172.26.0.1   36158.0  172.26.0.4   
110025  17/Jul/2020:23:36:11 +0200  172.26.0.1   50100.0  172.26.0.4   
110026  17/Jul/2020:23:36:12 +0200  172.26.0.1   50101.0  172.26.0.4   
110067  17/Jul/2020:23:36:19 +0200  172.26.0.1   50103.0  172.26.0.4   
110068  17/Jul/2020:23:36:19 +0200  172.26.0.1   50101.0  172.26.0.4   

        dst_port request_http_method  \
7491        80.0                 GET   
110025      80.0                POST   
110026      80.0                POST   
110067      80.0                 GET   
110068      80.0                POST   

                                     request_http_request  \
7491    /index.php?s=/Index/\x5Cthink\x5Capp/invokefun...   
110025          /blog/index.php/my-account/edit-password/   
110026          /blog/index.php/my-account/edit-password/   
110067  /blog/index.php/my-account/edit-profile/post.p...   
110068  /blog/index.php/my-account/edit-profile/post.p...   

       request_http_protocol  \
7491                HTTP/1.1   
110025              HTTP/1.1   
110026              HTTP/1.1   
110067              HTTP/1.1   
110068              HTTP/1.1   

                                       request_user_agent  \
7491    Mozilla/5.0 (Windows NT 10.0; Win64; x64) Appl...   
110025  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
110026  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
110067  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
110068  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   

                                          request_referer   request_host  \
7491                                                  NaN  test-site.com   
110025  http://test-site.com/blog/index.php/my-account...  test-site.com   
110026  http://test-site.com/blog/index.php/my-account...  test-site.com   
110067                                                NaN  test-site.com   
110068  http://test-site.com/blog/index.php/my-account...  test-site.com   

       request_origin                         request_cookie  \
7491              NaN                                    NaN   
110025            NaN  wordpress_test_cookie=WP+Cookie+check   
110026            NaN  wordpress_test_cookie=WP+Cookie+check   
110067            NaN  wordpress_test_cookie=WP+Cookie+check   
110068            NaN  wordpress_test_cookie=WP+Cookie+check   

                     request_content_type request_accept  \
7491                                  NaN            */*   
110025  application/x-www-form-urlencoded            NaN   
110026  application/x-www-form-urlencoded            NaN   
110067                                NaN            NaN   
110068  application/x-www-form-urlencoded            NaN   

       request_accept_language request_accept_encoding  request_do_not_track  \
7491                       NaN           deflate, gzip                   NaN   
110025                     NaN                     NaN                   NaN   
110026                     NaN                     NaN                   NaN   
110067                     NaN                     NaN                   NaN   
110068                     NaN                     NaN                   NaN   

       request_connection                                       request_body  \
7491                  NaN                                                NaN   
110025                NaN  username=%22%3E%3Cscript%3Ealert%281%29%3B%3C%...   
110026                NaN  username=%22+onMouseOver%3D%22alert%281%29%3B&...   
110067                NaN                                                NaN   
110068                NaN  username=wakefulViper3&password=GJLU7ABS&user-...   

       response_http_protocol  response_http_status_code  \
7491                 HTTP/1.1                      200.0   
110025               HTTP/1.1                      200.0   
110026               HTTP/1.1                      200.0   
110067             

## 88 - OS Command Injection

In [23]:
df_88 = df_capec[df_capec['88 - OS Command Injection'] == 1].loc[:, ['request_http_method',
                                                                        'request_http_request', 
                                                                        'request_body', 
                                                                        '88 - OS Command Injection']].reset_index(drop=True)
df_88.shape

(3074, 4)

In [24]:
df_88 = df_capec[df_capec['88 - OS Command Injection'] == 1]
df_88.shape

(3074, 39)

In [25]:
df_88.head()

timestamp      src_ip  src_port      dst_ip  \
7494    17/Jul/2020:20:08:52 +0200  172.26.0.1   36158.0  172.26.0.4   
257238  18/Jul/2020:04:08:39 +0200  172.26.0.1   63219.0  172.26.0.4   
257240  18/Jul/2020:04:08:39 +0200  172.26.0.1   63219.0  172.26.0.4   
257242  18/Jul/2020:04:08:40 +0200  172.26.0.1   63219.0  172.26.0.4   
257244  18/Jul/2020:04:08:40 +0200  172.26.0.1   63219.0  172.26.0.4   

        dst_port request_http_method  \
7494        80.0                 GET   
257238      80.0                POST   
257240      80.0                POST   
257242      80.0                POST   
257244      80.0                POST   

                                     request_http_request  \
7494    /index.php?s=/module/action/param1/${@die(sha1...   
257238          /blog/index.php/my-account/edit-password/   
257240          /blog/index.php/my-account/edit-password/   
257242          /blog/index.php/my-account/edit-password/   
257244          /blog/index.php/my-account/edit-password/   

       request_http_protocol  \
7494                HTTP/1.1   
257238              HTTP/1.1   
257240              HTTP/1.1   
257242              HTTP/1.1   
257244              HTTP/1.1   

                                       request_user_agent  \
7494    Mozilla/5.0 (Windows NT 10.0; Win64; x64) Appl...   
257238  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
257240  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
257242  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
257244  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   

                                          request_referer   request_host  \
7494                                                  NaN  test-site.com   
257238  http://test-site.com/blog/index.php/my-account...  test-site.com   
257240  http://test-site.com/blog/index.php/my-account...  test-site.com   
257242  http://test-site.com/blog/index.php/my-account...  test-site.com   
257244  http://test-site.com/blog/index.php/my-account...  test-site.com   

       request_origin                                     request_cookie  \
7494              NaN                                                NaN   
257238            NaN  comment_author_url_1aefbe2f76edd740f8e362f39da...   
257240            NaN  comment_author_url_1aefbe2f76edd740f8e362f39da...   
257242            NaN  comment_author_url_1aefbe2f76edd740f8e362f39da...   
257244            NaN  comment_author_url_1aefbe2f76edd740f8e362f39da...   

                     request_content_type request_accept  \
7494                                  NaN            */*   
257238  application/x-www-form-urlencoded            NaN   
257240  application/x-www-form-urlencoded            NaN   
257242  application/x-www-form-urlencoded            NaN   
257244  application/x-www-form-urlencoded            NaN   

       request_accept_language request_accept_encoding  request_do_not_track  \
7494                       NaN           deflate, gzip                   NaN   
257238                     NaN                     NaN                   NaN   
257240                     NaN                     NaN                   NaN   
257242                     NaN                     NaN                   NaN   
257244                     NaN                     NaN                   NaN   

       request_connection                                       request_body  \
7494                  NaN                                                NaN   
257238                NaN  password_current=aH2q*Rj%26sleep+15%26&passwor...   
257240                NaN  password_current=H#bO2%3Bsleep+15%3B&password_...   
257242                NaN  password_current=muRWD%22%26sleep+15%26%22&pas...   
257244                NaN  password_current=3X*Bam%22%3Bsleep+15%3B%22&pa...   

       response_http_protocol  response_http_status_code  \
7494                 HTTP/1.1                      200.0   
257238               HTTP/1.1                      200.0   
257240 

## 126 - Path Traversal


In [26]:
df_126 = df_capec[df_capec['126 - Path Traversal'] == 1].loc[:, ['request_http_method','request_http_request', 'request_body', '000 - Normal']].reset_index(drop=True)
df_126.shape

(17595, 4)

In [27]:
df_126 = df_capec[df_capec['126 - Path Traversal'] == 1]
df_126.shape


(17595, 39)

In [28]:
df_126.head()

timestamp      src_ip  src_port      dst_ip  dst_port  \
2968  17/Jul/2020:19:25:15 +0200  172.26.0.1   59862.0  172.26.0.4      80.0   
2970  17/Jul/2020:19:25:15 +0200  172.26.0.1   59862.0  172.26.0.4      80.0   
2973  17/Jul/2020:19:25:16 +0200  172.26.0.1   59862.0  172.26.0.4      80.0   
2975  17/Jul/2020:19:25:16 +0200  172.26.0.1   59862.0  172.26.0.4      80.0   
2976  17/Jul/2020:19:25:16 +0200  172.26.0.1   59862.0  172.26.0.4      80.0   

     request_http_method                     request_http_request  \
2968                 GET                            /.svn/entries   
2970                 GET                              /.git/index   
2973                 GET                       /blog/.svn/entries   
2975                 GET                         /blog/.git/index   
2976                 GET  /blog/index.php/my-account/.svn/entries   

     request_http_protocol                                 request_user_agent  \
2968              HTTP/1.1  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
2970              HTTP/1.1  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
2973              HTTP/1.1  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
2975              HTTP/1.1  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
2976              HTTP/1.1  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   

     request_referer   request_host request_origin  \
2968             NaN  test-site.com            NaN   
2970             NaN  test-site.com            NaN   
2973             NaN  test-site.com            NaN   
2975             NaN  test-site.com            NaN   
2976             NaN  test-site.com            NaN   

                                         request_cookie request_content_type  \
2968                                                NaN                  NaN   
2970                                                NaN                  NaN   
2973  wordpress_logged_in_1aefbe2f76edd740f8e362f39d...                  NaN   
2975  wordpress_logged_in_1aefbe2f76edd740f8e362f39d...                  NaN   
2976  wordpress_logged_in_1aefbe2f76edd740f8e362f39d...                  NaN   

     request_accept request_accept_language request_accept_encoding  \
2968            NaN                     NaN                     NaN   
2970            NaN                     NaN                     NaN   
2973            NaN                     NaN                     NaN   
2975            NaN                     NaN                     NaN   
2976            NaN                     NaN                     NaN   

      request_do_not_track request_connection request_body  \
2968                   NaN                NaN          NaN   
2970                   NaN                NaN          NaN   
2973                   NaN                NaN          NaN   
2975                   NaN                NaN          NaN   
2976                   NaN                NaN          NaN   

     response_http_protocol  response_http_status_code  \
2968               HTTP/1.1                      404.0   
2970               HTTP/1.1                      404.0   
2973               HTTP/1.1                      404.0   
2975               HTTP/1.1                      404.0   
2976               HTTP/1.1                      404.0   

     response_http_status_message  response_content_length  000 - Normal  \
2968                    Not Found                    276.0             0   
2970                    Not Found                    276.0             0   
2973                    Not Found                    276.0             0   
2975                    Not Found                    276.0             0   
2976                    Not Found                  58226.0             0   

      272 - Protocol Manipulation  242 - Code Injection  \
2968                            0                     0   
2970                            0                     0   
2973                            0                  

## 66 - SQL Injection

In [29]:
df_66 = df_capec[df_capec['66 - SQL Injection'] == 1].loc[:, ['request_http_method',
                                                              'request_http_request', 
                                                              'request_body', 
                                                              '66 - SQL Injection']].reset_index(drop=True)
df_66.shape

(248093, 4)

In [30]:
df_66 = df_capec[df_capec['66 - SQL Injection'] == 1]
df_66.shape

(248093, 39)

In [31]:
df_66.head()

timestamp      src_ip  src_port      dst_ip  dst_port  \
3121  17/Jul/2020:19:25:48 +0200  172.26.0.1   59867.0  172.26.0.4      80.0   
4255  17/Jul/2020:19:33:20 +0200  172.26.0.1   59968.0  172.26.0.4      80.0   
7159  17/Jul/2020:20:01:20 +0200  172.26.0.1   63063.0  172.26.0.4      80.0   
7758  17/Jul/2020:21:27:43 +0200  172.26.0.1   64594.0  172.26.0.4      80.0   
8188  17/Jul/2020:21:29:12 +0200  172.26.0.1   64609.0  172.26.0.4      80.0   

     request_http_method                               request_http_request  \
3121                 GET  /blog/wp-admin/load-styles.php?c=0&dir=ltr&loa...   
4255                 GET  /blog/wp-admin/load-styles.php?c=0&dir=ltr&loa...   
7159                 GET  /blog/wp-admin/load-styles.php?c=0&dir=ltr&loa...   
7758                 GET  /blog/wp-admin/load-styles.php?c=0&dir=ltr&loa...   
8188                 GET  /blog/wp-admin/load-styles.php?c=0&dir=ltr&loa...   

     request_http_protocol                                 request_user_agent  \
3121              HTTP/1.1  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
4255              HTTP/1.1  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
7159              HTTP/1.1  Mozilla/5.0 (X11; Linux x86_64; rv:73.0) Gecko...   
7758              HTTP/1.1  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
8188              HTTP/1.1  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   

                                        request_referer   request_host  \
3121             http://test-site.com/blog/wp-login.php  test-site.com   
4255                                                NaN  test-site.com   
7159  https://test-site.com/blog/wp-login.php?logged...  test-site.com   
7758  http://test-site.com/blog/wp-login.php?action=...  test-site.com   
8188                                                NaN  test-site.com   

     request_origin                         request_cookie  \
3121            NaN  wordpress_test_cookie=WP+Cookie+check   
4255            NaN  wordpress_test_cookie=WP+Cookie+check   
7159            NaN  wordpress_test_cookie=WP+Cookie+check   
7758            NaN  wordpress_test_cookie=WP+Cookie+check   
8188            NaN  wordpress_test_cookie=WP+Cookie+check   

     request_content_type      request_accept request_accept_language  \
3121                  NaN                 NaN                     NaN   
4255                  NaN                 NaN                     NaN   
7159                  NaN  text/css,*/*;q=0.1          en-US,en;q=0.5   
7758                  NaN                 NaN                     NaN   
8188                  NaN                 NaN                     NaN   

     request_accept_encoding  request_do_not_track request_connection  \
3121                     NaN                   NaN                NaN   
4255                     NaN                   NaN                NaN   
7159                     NaN                   1.0         keep-alive   
7758                     NaN                   NaN                NaN   
8188                     NaN                   NaN                NaN   

     request_body response_http_protocol  response_http_status_code  \
3121          NaN               HTTP/1.1                      200.0   
4255          NaN               HTTP/1.1                      200.0   
7159          NaN               HTTP/1.1                      200.0   
7758          NaN               HTTP/1.1                      200.0   
8188          NaN               HTTP/1.1                      200.0   

     response_http_status_message  response_content_length  000 - Normal  \
3121                           OK                      NaN             0   
4255                           OK                      NaN             0   
7159                           OK                      NaN             0   
7758                           OK                      NaN             0   
8188                           OK                      NaN             0   



## 16 - Dictionary-based Password Attack

In [32]:
df_16 = df_capec[df_capec['16 - Dictionary-based Password Attack'] == 1]
df_16.shape

(836, 39)

In [33]:
df_16.head()

timestamp      src_ip  src_port      dst_ip  dst_port  \
3122  17/Jul/2020:19:25:48 +0200  172.26.0.1   59863.0  172.26.0.4      80.0   
3351  17/Jul/2020:19:26:25 +0200  172.26.0.1   59870.0  172.26.0.4      80.0   
4565  17/Jul/2020:19:33:33 +0200  172.26.0.1   59973.0  172.26.0.4      80.0   
4730  17/Jul/2020:19:33:50 +0200  172.26.0.1   59975.0  172.26.0.4      80.0   
4731  17/Jul/2020:19:33:50 +0200  172.26.0.1   59975.0  172.26.0.4      80.0   

     request_http_method request_http_request request_http_protocol  \
3122                POST   /blog/wp-login.php              HTTP/1.1   
3351                POST   /blog/wp-login.php              HTTP/1.1   
4565                POST   /blog/wp-login.php              HTTP/1.1   
4730                POST   /blog/wp-login.php              HTTP/1.1   
4731                POST   /blog/wp-login.php              HTTP/1.1   

                                     request_user_agent  \
3122  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
3351  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
4565  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
4730  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
4731  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   

                                        request_referer   request_host  \
3122             http://test-site.com/blog/wp-login.php  test-site.com   
3351  http://test-site.com/blog/wp-login.php?loggedo...  test-site.com   
4565             http://test-site.com/blog/wp-login.php  test-site.com   
4730  http://test-site.com/blog/wp-login.php?reauth=...  test-site.com   
4731  http://test-site.com/blog/wp-login.php?reauth=...  test-site.com   

     request_origin                         request_cookie  \
3122            NaN  wordpress_test_cookie=WP+Cookie+check   
3351            NaN  wordpress_test_cookie=WP+Cookie+check   
4565            NaN  wordpress_test_cookie=WP+Cookie+check   
4730            NaN  wordpress_test_cookie=WP+Cookie+check   
4731            NaN  wordpress_test_cookie=WP+Cookie+check   

                   request_content_type request_accept  \
3122  application/x-www-form-urlencoded            NaN   
3351  application/x-www-form-urlencoded            NaN   
4565  application/x-www-form-urlencoded            NaN   
4730  application/x-www-form-urlencoded            NaN   
4731  application/x-www-form-urlencoded            NaN   

     request_accept_language request_accept_encoding  request_do_not_track  \
3122                     NaN                     NaN                   NaN   
3351                     NaN                     NaN                   NaN   
4565                     NaN                     NaN                   NaN   
4730                     NaN                     NaN                   NaN   
4731                     NaN                     NaN                   NaN   

     request_connection                                       request_body  \
3122                NaN  log=rafael&pwd=975cU&rememberme=forever&wp-sub...   
3351                NaN  log=thrilledCrackers6&pwd=hOKxTw_2&rememberme=...   
4565                NaN  log=innocentOwl1&pwd=41IPdtwh&rememberme=forev...   
4730                NaN  log=panickyEagle4&pwd=cXTkX&rememberme=forever...   
4731                NaN  log=trustingOatmeal8&pwd=qqIzM7vt&rememberme=f...   

     response_http_protocol  response_http_status_code  \
3122               HTTP/1.1                      200.0   
3351               HTTP/1.1                      200.0   
4565               HTTP/1.1                      200.0   
4730               HTTP/1.1                      200.0   
4731               HTTP/1.1                      200.0   

     response_http_status_message  response_content_length  000 - Normal  \
3122                           OK                   4215.0             0   
3351                           OK                   4209.0             0   
4565                           OK                   4209.0             0   
47

## 310 - Scanning for Vulnerable Software

In [34]:
df_310 = df_capec[df_capec['310 - Scanning for Vulnerable Software'] == 1]
df_310.shape

(2382, 39)

In [35]:
df_310.head()

timestamp      src_ip  src_port      dst_ip  dst_port  \
4738  17/Jul/2020:19:35:52 +0200  172.26.0.1   60004.0  172.26.0.4      80.0   
4741  17/Jul/2020:19:35:52 +0200  172.26.0.1   60012.0  172.26.0.4      80.0   
4742  17/Jul/2020:19:35:52 +0200  172.26.0.1   60006.0  172.26.0.4      80.0   
4747  17/Jul/2020:19:35:52 +0200  172.26.0.1   60013.0  172.26.0.4      80.0   
4748  17/Jul/2020:19:35:53 +0200  172.26.0.1   60008.0  172.26.0.4      80.0   

     request_http_method             request_http_request  \
4738                 GET        /nmaplowercheck1595007349   
4741                 GET        /nmaplowercheck1595007350   
4742                 GET                    /wp-login.php   
4747                 GET  /index.php/wp-json/wp/v2/posts/   
4748                 GET                 /crossdomain.xml   

     request_http_protocol                                 request_user_agent  \
4738              HTTP/1.1  Mozilla/5.0 (compatible; Nmap Scripting Engine...   
4741              HTTP/1.1  Mozilla/5.0 (compatible; Nmap Scripting Engine...   
4742              HTTP/1.1  Mozilla/5.0 (compatible; Nmap Scripting Engine...   
4747              HTTP/1.1  Mozilla/5.0 (compatible; Nmap Scripting Engine...   
4748              HTTP/1.1  Mozilla/5.0 (compatible; Nmap Scripting Engine...   

     request_referer                                request_host  \
4738             NaN  249.network5-175-62-15.static.network5.net   
4741             NaN  249.network5-175-62-15.static.network5.net   
4742             NaN  249.network5-175-62-15.static.network5.net   
4747             NaN  249.network5-175-62-15.static.network5.net   
4748             NaN  249.network5-175-62-15.static.network5.net   

     request_origin request_cookie request_content_type request_accept  \
4738            NaN            NaN                  NaN            NaN   
4741            NaN            NaN                  NaN            NaN   
4742            NaN            NaN                  NaN            NaN   
4747            NaN            NaN                  NaN            NaN   
4748            NaN            NaN                  NaN            NaN   

     request_accept_language request_accept_encoding  request_do_not_track  \
4738                     NaN                     NaN                   NaN   
4741                     NaN                     NaN                   NaN   
4742                     NaN                     NaN                   NaN   
4747                     NaN                     NaN                   NaN   
4748                     NaN                     NaN                   NaN   

     request_connection request_body response_http_protocol  \
4738              close          NaN               HTTP/1.1   
4741              close          NaN               HTTP/1.1   
4742              close          NaN               HTTP/1.1   
4747              close          NaN               HTTP/1.1   
4748              close          NaN               HTTP/1.1   

      response_http_status_code response_http_status_message  \
4738                      404.0                    Not Found   
4741                      404.0                    Not Found   
4742                      404.0                    Not Found   
4747                      404.0                    Not Found   
4748                      404.0                    Not Found   

      response_content_length  000 - Normal  272 - Protocol Manipulation  \
4738                    301.0             0                            0   
4741                    301.0             0                            0   
4742                    301.0             0                            0   
4747                    301.0             0                            0   
4748                    301.0             0                            0   

      242 - Code Injection  88 - OS Command Injection  126 - Path Traversal  \
4738                     0                          0            

## 153 - Input Data Manipulation

In [36]:
df_153 = df_capec[df_capec['153 - Input Data Manipulation'] == 1]
df_153.shape

(1387, 39)

In [37]:
df_153.head()

timestamp      src_ip  src_port      dst_ip  dst_port  \
19406  17/Jul/2020:21:48:16 +0200  172.26.0.1   64765.0  172.26.0.4      80.0   
19411  17/Jul/2020:21:48:17 +0200  172.26.0.1   64763.0  172.26.0.4      80.0   
19413  17/Jul/2020:21:48:17 +0200  172.26.0.1   64765.0  172.26.0.4      80.0   
19415  17/Jul/2020:21:48:17 +0200  172.26.0.1   64765.0  172.26.0.4      80.0   
19416  17/Jul/2020:21:48:17 +0200  172.26.0.1   64763.0  172.26.0.4      80.0   

      request_http_method                       request_http_request  \
19406                POST  /blog/index.php/my-account/edit-password/   
19411                POST  /blog/index.php/my-account/edit-password/   
19413                POST  /blog/index.php/my-account/edit-password/   
19415                POST  /blog/index.php/my-account/edit-password/   
19416                POST  /blog/index.php/my-account/edit-password/   

      request_http_protocol  \
19406              HTTP/1.1   
19411              HTTP/1.1   
19413              HTTP/1.1   
19415              HTTP/1.1   
19416              HTTP/1.1   

                                      request_user_agent  \
19406  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
19411  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
19413  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
19415  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   
19416  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...   

                                         request_referer   request_host  \
19406  http://test-site.com/blog/index.php/my-account...  test-site.com   
19411  http://test-site.com/blog/index.php/my-account...  test-site.com   
19413  http://test-site.com/blog/index.php/my-account...  test-site.com   
19415  http://test-site.com/blog/index.php/my-account...  test-site.com   
19416  http://test-site.com/blog/index.php/my-account...  test-site.com   

      request_origin                         request_cookie  \
19406            NaN  wordpress_test_cookie=WP+Cookie+check   
19411            NaN  wordpress_test_cookie=WP+Cookie+check   
19413            NaN  wordpress_test_cookie=WP+Cookie+check   
19415            NaN  wordpress_test_cookie=WP+Cookie+check   
19416            NaN  wordpress_test_cookie=WP+Cookie+check   

      request_content_type request_accept request_accept_language  \
19406          /etc/passwd            NaN                     NaN   
19411           etc/passwd            NaN                     NaN   
19413                  c:/            NaN                     NaN   
19415                    /            NaN                     NaN   
19416                  c:\            NaN                     NaN   

      request_accept_encoding  request_do_not_track request_connection  \
19406                     NaN                   NaN                NaN   
19411                     NaN                   NaN                NaN   
19413                     NaN                   NaN                NaN   
19415                     NaN                   NaN                NaN   
19416                     NaN                   NaN                NaN   

                                            request_body  \
19406  password_current=76U8B7&password_1=76U8B7&pass...   
19411  password_current=ltHgBOO&password_1=ltHgBOO&pa...   
19413  password_current=5uSt9Xd&password_1=5uSt9Xd&pa...   
19415  password_current=J875f&password_1=J875f&passwo...   
19416  password_current=C&rclFh&password_1=C&rclFh&pa...   

      response_http_protocol  response_http_status_code  \
19406               HTTP/1.1                      200.0   
19411               HTTP/1.1                      200.0   
19413               HTTP/1.1                      200.0   
19415               HTTP/1.1                      200.0   
19416               HTTP/1.1                      200.0   

      response_http_status_message  response_content_length  000 - Normal  \
19406                           OK                  57821.0             0

## 274 - HTTP Verb Tampering

In [38]:
df_274 = df_capec[df_capec['274 - HTTP Verb Tampering'] == 1]
df_274.shape

(4047, 39)

In [39]:
df_274.head()

timestamp      src_ip  src_port      dst_ip  \
534146  19/Jul/2020:11:41:49 +0200  172.26.0.1   54485.0  172.26.0.4   
534156  19/Jul/2020:11:41:49 +0200  172.26.0.1   54488.0  172.26.0.4   
534163  19/Jul/2020:11:41:50 +0200  172.26.0.1   54486.0  172.26.0.4   
534170  19/Jul/2020:11:41:51 +0200  172.26.0.1   54492.0  172.26.0.4   
534177  19/Jul/2020:11:41:51 +0200  172.26.0.1   54493.0  172.26.0.4   

        dst_port request_http_method       request_http_request  \
534146      80.0               TRACK                     /O8AQS   
534156      80.0               TRACK                 /blogYDQJC   
534163      80.0               TRACK       /blog/index.php0NEKM   
534170      80.0               TRACK       /blog/index.php87RYI   
534177      80.0               TRACK  /blog/index.php/2020SS8Q8   

       request_http_protocol request_user_agent request_referer  \
534146              HTTP/1.1                NaN             NaN   
534156              HTTP/1.1                NaN             NaN   
534163              HTTP/1.1                NaN             NaN   
534170              HTTP/1.1                NaN             NaN   
534177              HTTP/1.1                NaN             NaN   

         request_host request_origin  \
534146  test-site.com            NaN   
534156  test-site.com            NaN   
534163  test-site.com            NaN   
534170  test-site.com            NaN   
534177  test-site.com            NaN   

                                           request_cookie  \
534146                                                NaN   
534156                                                NaN   
534163  wordpress_logged_in_1aefbe2f76edd740f8e362f39d...   
534170  wordpress_logged_in_1aefbe2f76edd740f8e362f39d...   
534177  wordpress_logged_in_1aefbe2f76edd740f8e362f39d...   

       request_content_type request_accept request_accept_language  \
534146                  NaN            NaN                     NaN   
534156                  NaN            NaN                     NaN   
534163                  NaN            NaN                     NaN   
534170                  NaN            NaN                     NaN   
534177                  NaN            NaN                     NaN   

       request_accept_encoding  request_do_not_track request_connection  \
534146                     NaN                   NaN                NaN   
534156                     NaN                   NaN                NaN   
534163                     NaN                   NaN                NaN   
534170                     NaN                   NaN                NaN   
534177                     NaN                   NaN                NaN   

       request_body response_http_protocol  response_http_status_code  \
534146          NaN               HTTP/1.1                      501.0   
534156          NaN               HTTP/1.1                      501.0   
534163          NaN               HTTP/1.1                      501.0   
534170          NaN               HTTP/1.1                      501.0   
534177          NaN               HTTP/1.1                      404.0   

       response_http_status_message  response_content_length  000 - Normal  \
534146              Not Implemented                    285.0             0   
534156              Not Implemented                    285.0             0   
534163              Not Implemented                    285.0             0   
534170              Not Implemented                    285.0             0   
534177                    Not Found                  58226.0             0   

        272 - Protocol Manipulation  242 - Code Injection  \
534146                            0                     0   
534156                            0                     0   
534163                            0                     0   
534170                            0                     0   
534177                            0                     0   

        88 - OS Command Injecti

## 194 - Fake the Source of Data

In [40]:
df_194 = df_capec[df_capec['194 - Fake the Source of Data'] == 1]
df_194.shape

(55982, 39)

In [41]:
df_194.head()

timestamp      src_ip  src_port      dst_ip  dst_port  \
44501  17/Jul/2020:22:24:07 +0200  172.26.0.1   65225.0  172.26.0.4      80.0   
44502  17/Jul/2020:22:24:07 +0200  172.26.0.1   65227.0  172.26.0.4      80.0   
44504  17/Jul/2020:22:24:07 +0200  172.26.0.1   65228.0  172.26.0.4      80.0   
44505  17/Jul/2020:22:24:07 +0200  172.26.0.1   65227.0  172.26.0.4      80.0   
44506  17/Jul/2020:22:24:07 +0200  172.26.0.1   65227.0  172.26.0.4      80.0   

      request_http_method                            request_http_request  \
44501                 GET                 /http%3A%2F%2Fwww.google.com%2F   
44502                 GET       /http%3A%2F%2Fwww.google.com%2F/index.php   
44504                 GET            /http%3A%2F%2Fwww.google.com%3A80%2F   
44505                 GET  /http%3A%2F%2Fwww.google.com%3A80%2F/index.php   
44506                 GET                    /http%3A%2F%2Fwww.google.com   

      request_http_protocol  \
44501              HTTP/1.1   
44502              HTTP/1.1   
44504              HTTP/1.1   
44505              HTTP/1.1   
44506              HTTP/1.1   

                                      request_user_agent request_referer  \
44501  Mozilla/5.0 (X11; Linux x86_64; rv:73.0) Gecko...             NaN   
44502  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...             NaN   
44504  Mozilla/5.0 (X11; Linux x86_64; rv:73.0) Gecko...             NaN   
44505  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...             NaN   
44506  Mozilla/5.0 (X11; Linux x86_64; rv:73.0) Gecko...             NaN   

        request_host request_origin  \
44501  test-site.com            NaN   
44502  test-site.com            NaN   
44504  test-site.com            NaN   
44505  test-site.com            NaN   
44506  test-site.com            NaN   

                                          request_cookie request_content_type  \
44501                                                NaN                  NaN   
44502  wordpress_logged_in_1aefbe2f76edd740f8e362f39d...                  NaN   
44504                                                NaN                  NaN   
44505  wordpress_logged_in_1aefbe2f76edd740f8e362f39d...                  NaN   
44506                                                NaN                  NaN   

                                          request_accept  \
44501  text/html,application/xhtml+xml,application/xm...   
44502                                                NaN   
44504  text/html,application/xhtml+xml,application/xm...   
44505                                                NaN   
44506  text/html,application/xhtml+xml,application/xm...   

      request_accept_language request_accept_encoding  request_do_not_track  \
44501          en-US,en;q=0.5                     NaN                   NaN   
44502                     NaN                     NaN                   NaN   
44504          en-US,en;q=0.5                     NaN                   NaN   
44505                     NaN                     NaN                   NaN   
44506          en-US,en;q=0.5                     NaN                   NaN   

      request_connection request_body response_http_protocol  \
44501         keep-alive          NaN               HTTP/1.1   
44502                NaN          NaN               HTTP/1.1   
44504         keep-alive          NaN               HTTP/1.1   
44505                NaN          NaN               HTTP/1.1   
44506         keep-alive          NaN               HTTP/1.1   

       response_http_status_code response_http_status_message  \
44501                      404.0                    Not Found   
44502                      404.0                    Not Found   
44504                      404.0                    Not Found   
44505                      404.0                    Not Found   
44506                      404.0                    Not Found   

       response_content_length  000 - Normal  272 - Protocol Manipulation  \
44501              

## 34 - HTTP Response Splitting

In [42]:
df_34 = df_capec[df_capec['34 - HTTP Response Splitting'] == 1].loc[:, ['request_http_method',
                                                                        'request_http_request', 
                                                                        'request_body', 
                                                                        '34 - HTTP Response Splitting']].reset_index(drop=True)
df_34.shape

(19134, 4)

In [43]:
df_34 = df_capec[df_capec['34 - HTTP Response Splitting'] == 1]
df_34.shape

(19134, 39)

In [44]:
df_34.head()

timestamp      src_ip  src_port      dst_ip  \
242072  18/Jul/2020:03:22:32 +0200  172.26.0.1   62915.0  172.26.0.4   
242073  18/Jul/2020:03:22:32 +0200  172.26.0.1   62914.0  172.26.0.4   
242096  18/Jul/2020:03:22:34 +0200  172.26.0.1   62915.0  172.26.0.4   
242097  18/Jul/2020:03:22:34 +0200  172.26.0.1   62914.0  172.26.0.4   
242121  18/Jul/2020:03:22:37 +0200  172.26.0.1   62915.0  172.26.0.4   

        dst_port request_http_method  \
242072      80.0                 GET   
242073      80.0                 GET   
242096      80.0                 GET   
242097      80.0                 GET   
242121      80.0                 GET   

                            request_http_request request_http_protocol  \
242072            /blog%0Acat+%2Fetc%2Fpasswd%0A              HTTP/1.1   
242073  /blog%0Acat+%2Fetc%2Fpasswd%0A/index.php              HTTP/1.1   
242096                       /blog%0Asleep+15%0A              HTTP/1.1   
242097             /blog%0Asleep+15%0A/index.php              HTTP/1.1   
242121  /blog%0Acat+%2Fetc%2Fpasswd%0A/index.php              HTTP/1.1   

                                       request_user_agent request_referer  \
242072  Mozilla/5.0 (X11; Linux x86_64; rv:73.0) Gecko...             NaN   
242073  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...             NaN   
242096  Mozilla/5.0 (X11; Linux x86_64; rv:73.0) Gecko...             NaN   
242097  Mozilla/5.0 (X11; Linux x86_64; rv:70.0) Gecko...             NaN   
242121  Mozilla/5.0 (X11; Linux x86_64; rv:73.0) Gecko...             NaN   

         request_host request_origin  \
242072  test-site.com            NaN   
242073  test-site.com            NaN   
242096  test-site.com            NaN   
242097  test-site.com            NaN   
242121  test-site.com            NaN   

                                           request_cookie  \
242072                                                NaN   
242073  wordpress_logged_in_1aefbe2f76edd740f8e362f39d...   
242096                                                NaN   
242097  wordpress_logged_in_1aefbe2f76edd740f8e362f39d...   
242121                                                NaN   

       request_content_type  \
242072                  NaN   
242073                  NaN   
242096                  NaN   
242097                  NaN   
242121                  NaN   

                                           request_accept  \
242072  text/html,application/xhtml+xml,application/xm...   
242073                                                NaN   
242096  text/html,application/xhtml+xml,application/xm...   
242097                                                NaN   
242121  text/html,application/xhtml+xml,application/xm...   

       request_accept_language request_accept_encoding  request_do_not_track  \
242072          en-US,en;q=0.5                     NaN                   NaN   
242073                     NaN                     NaN                   NaN   
242096          en-US,en;q=0.5                     NaN                   NaN   
242097                     NaN                     NaN                   NaN   
242121          en-US,en;q=0.5                     NaN                   NaN   

       request_connection request_body response_http_protocol  \
242072         keep-alive          NaN               HTTP/1.1   
242073                NaN          NaN               HTTP/1.1   
242096         keep-alive          NaN               HTTP/1.1   
242097                NaN          NaN               HTTP/1.1   
242121         keep-alive          NaN               HTTP/1.1   

        response_http_status_code response_http_status_message  \
242072                      404.0                    Not Found   
242073                      404.0                    Not Found   
242096                      404.0                    Not Found   
242097                      404.0                    Not Found   
242121                      404.0                    Not Found   

        resp

## Note on data
- Có 2 trường attack ko có tấn công trong bộ dataset multiclass:
    + 33 - HTTP Request Smuggling 
    + 248 - Command Injection 
Lý do có thể vì hai trường attack này giao với các trường attack khác (multilabel)
- Tổng cộng các loại tấn công (đã bỏ 2 loại attack bên trên): 12 (bao gồm normal):
    1. 000 - Normal: 226509
    2. 272 - Protocol Manipulation: 6924 
    3. 242 - Code Injection: 13792
    4. 88 - OS Command Injection: 3074
    5. 126 - Path Traversal: 17595 
    6. 66 - SQL Injection: 248093
    7. 16 - Dictionary-based Password Attack: 836
    8. 310 - Scanning for Vulnerable Software: 2382 
    9. 153 - Input Data Manipulation: 1387
    10. 274 - HTTP Verb Tampering: 4047
    11. 194 - Fake the Source of Data: 55982
    12. 34 - HTTP Response Splitting: 19134
- Mỗi sample có rất nhiều trường dữ liệu, tuy nhiên chỉ có 1 số attribute giúp classify web attacking
- Thử nghiệm với concat 2 trường dữ liệu: request_http_request và request_body (với method POST)
- Thử nghiệm với một vài kiểu tấn công, và để lại một vài kiểu tấn công có cùng variance với bộ dataset để nghiên cứu Transfer Learning
- Các kiểu tấn công sử dụng trong dataset: 
    + 272 - Protocol Manipulation
    + 242 - Code Injection
    + 126 - Path Traversal
    + 66 - SQL Injection
    + 310 - Scanning for Vulnerable Software
    + 153 - Input Data Manipulation
    + 194 - Fake the Source of Data
    + 34 - HTTP Response Splitting
- Kiểu tấn công dành cho Transfer Learning: 88-OS Command Injection
- Hiện tại đang chưa tích hợp bộ dataset CSIC 2010-2012 (Bao gồm XSS attack và SQLi attack)

## Prepare and format data

In [45]:
# Concat http_method with http_request and http_content

In [46]:
def create_text(method, request, body):
    if method == 'GET':
        return str(method) + ' ' + str(request)
    return str(method) + ' ' + str(request) + ' ' + str(body)

df_capec['text'] = df_capec.apply(lambda x: create_text(x['request_http_method'], 
                                                        x['request_http_request'], 
                                                        x['request_body']), axis = 1)
df_capec.head()

timestamp      src_ip  src_port      dst_ip  dst_port  \
1  17/Jul/2020:12:23:41 +0100  172.26.0.1   55897.0  172.26.0.4      80.0   
2  17/Jul/2020:12:23:51 +0100  172.26.0.1   55901.0  172.26.0.4      80.0   
4  17/Jul/2020:12:23:58 +0100  172.26.0.1   55903.0  172.26.0.4      80.0   
5  17/Jul/2020:12:24:07 +0100  172.26.0.1   55910.0  172.26.0.4      80.0   
6  17/Jul/2020:12:24:15 +0100  172.26.0.1   55911.0  172.26.0.4      80.0   

  request_http_method                               request_http_request  \
1                 GET  /blog/index.php/2020/04/04/voluptatum-reprehen...   
2                 GET                               /blog/xmlrpc.php?rsd   
4                 GET  /blog/index.php/2020/04/04/nihil-tenetur-et-ad...   
5                 GET  /blog/index.php/2020/04/04/explicabo-qui-fuga-...   
6                 GET  /blog/index.php/2020/04/04/explicabo-qui-fuga-...   

  request_http_protocol                                 request_user_agent  \
1              HTTP/1.1  Mozilla/5.0 (Macintosh; Intel Mac OS X 10.6; r...   
2              HTTP/1.1  Mozilla/4.0 (compatible; MSIE 6.0; Windows NT ...   
4              HTTP/1.1  Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_2...   
5              HTTP/1.1  Mozilla/5.0 (Windows NT 10.0; Win64; x64) Appl...   
6              HTTP/1.1  Mozilla/5.0 (Windows NT 5.1) AppleWebKit/537.3...   

  request_referer   request_host request_origin request_cookie  \
1             NaN  test-site.com            NaN            NaN   
2             NaN  test-site.com            NaN            NaN   
4             NaN  test-site.com            NaN            NaN   
5             NaN  test-site.com            NaN            NaN   
6             NaN  test-site.com            NaN            NaN   

  request_content_type request_accept request_accept_language  \
1                  NaN            */*                     NaN   
2                  NaN            */*                     NaN   
4                  NaN            */*                     NaN   
5                  NaN            */*                     NaN   
6                  NaN            */*                     NaN   

  request_accept_encoding  request_do_not_track request_connection  \
1           gzip, deflate                   NaN         keep-alive   
2           gzip, deflate                   NaN         keep-alive   
4           gzip, deflate                   NaN         keep-alive   
5           gzip, deflate                   NaN         keep-alive   
6           gzip, deflate                   NaN         keep-alive   

  request_body response_http_protocol  response_http_status_code  \
1          NaN               HTTP/1.1                      200.0   
2          NaN               HTTP/1.1                      200.0   
4          NaN               HTTP/1.1                      200.0   
5          NaN               HTTP/1.1                      200.0   
6          NaN               HTTP/1.1                      200.0   

  response_http_status_message  response_content_length  000 - Normal  \
1                           OK                  20862.0             1   
2                           OK                    317.0             1   
4                           OK                  20863.0             1   
5                           OK                  21062.0             1   
6                           OK                  21061.0             1   

   272 - Protocol Manipulation  242 - Code Injection  \
1                            0                     0   
2                            0                     0   
4                            0                     0   
5                            0                     0   
6                            0                     0   

   88 - OS Command Injection  126 - Path Traversal  66 - SQL Injection  \
1                          0                     0                   0   
2                          0                     0                   0   
4              

In [47]:
# Mapping values for classification

def count_duplicate(a0,a1,a2,a3,a4,a5,a6,a7,a8,a9,a10,a11,a12,a13):
    if a0 == 1:
        return '000 - Normal'
    elif a1 == 1:
        return '272 - Protocol Manipulation'
    elif a2 == 1:
        return '242 - Code Injection'
    elif a3 == 1:
        return '88 - OS Command Injection'
    elif a4 == 1:
        return '126 - Path Traversal'
    elif a5 == 1:
        return '66 - SQL Injection'
    elif a6 == 1:
        return '16 - Dictionary-based Password Attack'
    elif a7 == 1:
        return '310 - Scanning for Vulnerable Software'
    elif a8 == 1:
        return '153 - Input Data Manipulation'
    elif a9 == 1:
        return '248 - Command Injection'
    elif a10 == 1:
        return '274 - HTTP Verb Tampering'
    elif a11 == 1:
        return '194 - Fake the Source of Data'
    elif a12 == 1:
        return '34 - HTTP Response Splitting'
    elif a13 == 1:
        return '33 - HTTP Request Smuggling'
    return 'Error'

df_capec['label'] = df_capec.apply(lambda x: count_duplicate(x['000 - Normal'],
                                                             x['272 - Protocol Manipulation'], 
                                                             x['242 - Code Injection'],
                                                             x['88 - OS Command Injection'],
                                                             x['126 - Path Traversal'],
                                                             x['66 - SQL Injection'],
                                                             x['16 - Dictionary-based Password Attack'],
                                                             x['310 - Scanning for Vulnerable Software'],
                                                             x['153 - Input Data Manipulation'],
                                                             x['248 - Command Injection'],
                                                             x['274 - HTTP Verb Tampering'],
                                                             x['194 - Fake the Source of Data'],
                                                             x['34 - HTTP Response Splitting'],
                                                             x['33 - HTTP Request Smuggling']), axis = 1)
df_capec.head()

timestamp      src_ip  src_port      dst_ip  dst_port  \
1  17/Jul/2020:12:23:41 +0100  172.26.0.1   55897.0  172.26.0.4      80.0   
2  17/Jul/2020:12:23:51 +0100  172.26.0.1   55901.0  172.26.0.4      80.0   
4  17/Jul/2020:12:23:58 +0100  172.26.0.1   55903.0  172.26.0.4      80.0   
5  17/Jul/2020:12:24:07 +0100  172.26.0.1   55910.0  172.26.0.4      80.0   
6  17/Jul/2020:12:24:15 +0100  172.26.0.1   55911.0  172.26.0.4      80.0   

  request_http_method                               request_http_request  \
1                 GET  /blog/index.php/2020/04/04/voluptatum-reprehen...   
2                 GET                               /blog/xmlrpc.php?rsd   
4                 GET  /blog/index.php/2020/04/04/nihil-tenetur-et-ad...   
5                 GET  /blog/index.php/2020/04/04/explicabo-qui-fuga-...   
6                 GET  /blog/index.php/2020/04/04/explicabo-qui-fuga-...   

  request_http_protocol                                 request_user_agent  \
1              HTTP/1.1  Mozilla/5.0 (Macintosh; Intel Mac OS X 10.6; r...   
2              HTTP/1.1  Mozilla/4.0 (compatible; MSIE 6.0; Windows NT ...   
4              HTTP/1.1  Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_2...   
5              HTTP/1.1  Mozilla/5.0 (Windows NT 10.0; Win64; x64) Appl...   
6              HTTP/1.1  Mozilla/5.0 (Windows NT 5.1) AppleWebKit/537.3...   

  request_referer   request_host request_origin request_cookie  \
1             NaN  test-site.com            NaN            NaN   
2             NaN  test-site.com            NaN            NaN   
4             NaN  test-site.com            NaN            NaN   
5             NaN  test-site.com            NaN            NaN   
6             NaN  test-site.com            NaN            NaN   

  request_content_type request_accept request_accept_language  \
1                  NaN            */*                     NaN   
2                  NaN            */*                     NaN   
4                  NaN            */*                     NaN   
5                  NaN            */*                     NaN   
6                  NaN            */*                     NaN   

  request_accept_encoding  request_do_not_track request_connection  \
1           gzip, deflate                   NaN         keep-alive   
2           gzip, deflate                   NaN         keep-alive   
4           gzip, deflate                   NaN         keep-alive   
5           gzip, deflate                   NaN         keep-alive   
6           gzip, deflate                   NaN         keep-alive   

  request_body response_http_protocol  response_http_status_code  \
1          NaN               HTTP/1.1                      200.0   
2          NaN               HTTP/1.1                      200.0   
4          NaN               HTTP/1.1                      200.0   
5          NaN               HTTP/1.1                      200.0   
6          NaN               HTTP/1.1                      200.0   

  response_http_status_message  response_content_length  000 - Normal  \
1                           OK                  20862.0             1   
2                           OK                    317.0             1   
4                           OK                  20863.0             1   
5                           OK                  21062.0             1   
6                           OK                  21061.0             1   

   272 - Protocol Manipulation  242 - Code Injection  \
1                            0                     0   
2                            0                     0   
4                            0                     0   
5                            0                     0   
6                            0                     0   

   88 - OS Command Injection  126 - Path Traversal  66 - SQL Injection  \
1                          0                     0                   0   
2                          0                     0                   0   
4              

In [48]:
df_training = df_capec[(df_capec['label'] != '33 - HTTP Request Smuggling')
                       & (df_capec['label'] != '248 - Command Injection') 
                       & (df_capec['label'] != '16 - Dictionary-based Password Attack') 
                       & (df_capec['label'] != '274 - HTTP Verb Tampering') 
                       & (df_capec['label'] != '88 - OS Command Injection')]
df_training.shape

(591798, 41)

In [49]:
df_capec.shape

(599755, 41)

In [50]:
df_training.label.unique()

array(['000 - Normal', '126 - Path Traversal', '66 - SQL Injection',
       '272 - Protocol Manipulation',
       '310 - Scanning for Vulnerable Software', '242 - Code Injection',
       '153 - Input Data Manipulation', '194 - Fake the Source of Data',
       '34 - HTTP Response Splitting'], dtype=object)

In [51]:
df_final = df_training.loc[:, ['text','label']].reset_index(drop=True)
df_final.head()

text         label
0  GET /blog/index.php/2020/04/04/voluptatum-repr...  000 - Normal
1                           GET /blog/xmlrpc.php?rsd  000 - Normal
2  GET /blog/index.php/2020/04/04/nihil-tenetur-e...  000 - Normal
3  GET /blog/index.php/2020/04/04/explicabo-qui-f...  000 - Normal
4  GET /blog/index.php/2020/04/04/explicabo-qui-f...  000 - Normal

In [52]:
# Save to csv
df_final.to_csv('Dataset/dataset_capec_test.csv', encoding='utf-8', index=False)